pkg/relayer/miner/miner_test.go

```diff
@@ -1,10 +1,256 @@
+	hasher := newHasher()
+	_, err := hasher.Write(relayBz)
+	require.NoError(t, err)
+	return hasher.Sum(nil)
+}
+
+func TestSimpleMineMockRelays(t *testing.T) {
+	t.Skip("this test is intended to be run manually as a utility to generate relay fixtures for testing")
+
+	ctx := context.Background()
+
+	minedBytesObs := mineRelayFixturesForDifficulty(
+		t,
+		16,
+		testDifficulty,
+		5,
+		miner.DefaultRelayHasher,
+	)
+	minedBytesObserver := minedBytesObs.Subscribe(ctx)
+
+	for minedRelay := range minedBytesObserver.Ch() {
+		minedRelayBz, err := minedRelay.Marshal()
+		require.NoError(t, err)
+
+		t.Logf("%x", minedRelayBz)
+	}
+}
+
+// mineRelayFixturesForDifficulty is a single-threaded utility for generating
+// relay fixtures for testing. It returns an observable of mined relays which
+// are published as they are mined, where difficulty is the difficulty threshold
+// in bytes. Each relay fixture is populated with a randomized signature of
+// randLength length. Relay fixtures are then hashed using the hasher returned
+// from newHasher and checked against the difficulty threshold. If the relay
+// meets the difficulty threshold, it is published to the returned observable.
+// It stops mining & publishing once limit number of relay fixtures have been
+// published.
+// It is not intended to be used **in/by** any tests but rather is persisted to
+// aid in re-generation of relay fixtures should the test requirements change.
+func mineRelayFixturesForDifficulty(
+	t *testing.T,
+	randLength int,
+	difficulty int,
+	limit int,
+	newHasher func() hash.Hash,
+) observable.Observable[*relayer.MinedRelay] {
+	t.Helper()
+
+	var (
+		mined                      = 0
+		attempted                  = 0
+		randBzObs, randBzPublishCh = channel.NewObservable[*relayer.MinedRelay]()
+	)
+
+	go func() {
+		for {
+			if mined >= limit {
+				break
+			}
+
+			randBz := make([]byte, randLength)
+			n, err := rand.Read(randBz)
+			require.NoError(t, err)
+			require.Equal(t, randLength, n)
+
+			// Populate a relay with the minimally sufficient randomized data.
+			relay := servicetypes.Relay{
+				Req: &servicetypes.RelayRequest{
+					Meta: &servicetypes.RelayRequestMetadata{
+						Signature: randBz,
+					},
+					Payload: nil,
+				},
+				Res: nil,
+			}
+
+			// TODO_BLOCKER: use canonical codec.
+			relayBz, err := relay.Marshal()
+			require.NoError(t, err)
+
+			// Hash relay bytes
+			relayHashBz := hashRelay(t, newHasher, relayBz)
+
+			// Check difficulty & publish.
+			if !protocol.BytesDifficultyGreaterThan(relayHashBz, difficulty) {
+				randBzPublishCh <- &relayer.MinedRelay{
+					Relay: relay,
+					Bytes: relayBz,
+					Hash:  relayHashBz,
+				}
+				mined++
+			}
+			attempted++
+
+			// Log occasionally; signal liveness/progress.
+			if attempted%100000 == 0 {
+				t.Logf("attempted: %d, mined: %d", attempted, mined)
+			}
+		}
+		close(randBzPublishCh)
+	}()
+
+	return randBzObs
```

Comment:
I ❤️  everything about this function

Reply:
I :heart: that you :heart: that, I hope you like the next iteration as well and/or apologies for the rug-pull :sweat_smile:.